crates/cxx-qt-gen/src/generator/naming/property.rs

```diff
@@ -2,32 +2,107 @@
+    pub fn from_flag_with_auto_fn(state: &FlagState, auto_fn: impl Fn() -> Name) -> Self {
+        match state {
+            FlagState::Auto => Self::Auto(auto_fn()),
+            FlagState::Custom(ident) => Self::Custom(Name::new(ident.clone())),
+        }
+    }
+}
+
 /// Names for parts of a Q_PROPERTY
 pub struct QPropertyNames {
     pub name: Name,
-    pub getter: Name,
+    pub getter: NameState,
     pub getter_wrapper: Name,
-    pub setter: Name,
-    pub setter_wrapper: Name,
-    pub notify: Name,
+    pub setter: Option<NameState>,
+    pub setter_wrapper: Option<Name>,
+    pub notify: Option<NameState>,
 }
 
 impl From<&ParsedQProperty> for QPropertyNames {
     fn from(property: &ParsedQProperty) -> Self {
         let property_name = property_name_from_rust_name(property.ident.clone());
-        let getter = getter_name_from_property(&property_name);
-        let setter = setter_name_from_property(&property_name);
+
+        let flags = &property.flags;
+
+        // let getter = match &flags.read {
+        //     FlagState::Auto => getter_name_from_property(&property_name),
+        //     FlagState::Custom(ident) => Name::new(ident.clone()),
+        // };
+
+        // let setter = match &flags.write {
+        //     Some(state) => match state {
+        //         FlagState::Auto => Some(setter_name_from_property(&property_name)),
+        //         FlagState::Custom(ident) => Some(Name::new(ident.clone())),
+        //     },
+        //     None => None,
+        // };
+
+        // let notify = match &flags.notify {
+        //     Some(state) => match state {
+        //         FlagState::Auto => Some(notify_name_from_property(&property_name)),
+        //         FlagState::Custom(ident) => Some(Name::new(ident.clone())),
+        //     },
+        //     None => None,
+        // };
+
+        let getter = NameState::from_flag_with_auto_fn(&flags.read, || {
+            getter_name_from_property(&property_name)
+        });
+
+        let setter = flags.write.as_ref().map(|setter| {
+            NameState::from_flag_with_auto_fn(setter, || setter_name_from_property(&property_name))
+        });
+
+        let notify = flags.notify.as_ref().map(|notify| {
+            NameState::from_flag_with_auto_fn(notify, || notify_name_from_property(&property_name))
+        });
+
+        // let setter_wrapper = setter
+        //     .as_ref()
+        //     .map(|state| wrapper_name_from_function_name(state));
+
+        let setter_wrapper = if let Some(NameState::Auto(ref setter)) = setter {
+            Some(wrapper_name_from_function_name(setter))
+        } else {
+            None
+        };
+
         Self {
             getter_wrapper: wrapper_name_from_function_name(&getter),
             getter,
-            setter_wrapper: wrapper_name_from_function_name(&setter),
+            setter_wrapper,
             setter,
-            notify: notify_name_from_property(&property_name),
+            notify,
```

Comment:
something we need to figure out is how to read the cxx_name/rust_name correctly of the custom functions, eg...

i believe that this is will pick up the wrong notify name in the following case

```rust
extern "RustQt" {
   // here we should always use the rust name for custom attribute values
   #[qproperty(T, name, read = prop_getter, write = prop_setter, notify = prop_changed]
   #[qobject]
   type MyObject = super::MyObjectRust;

   #[cxx_name = "getProp"]
   fn prop_getter(self: &MyObject) -> T;
   #[cxx_name = "setProp"]
   fn prop_setter(self: Pin<&mut MyObject>, value: T);   

   #[qsignal]
   #[cxx_name = "propChanged"]
   fn prop_changed(self: Pin<&mut MyObject);
}
```

In these cases our Rust side would be correct but our C++ Q_PROPERTY would be incorrect as it would the custom name from the qproperty instead of the cxx_name. (there is also the reverse fun for the rust_name).

So when the flag is Custom we need to somehow lookup the given name in i think TypeNames, eg with `TypeNames::lookup` which takes an Ident and gives you a Name (which would have the cxx and rust name from the other blocks.

`generate_cpp_properties` already has `type_names: &TypeNames,` so i think it's just a matter of passing that into QPropertyNames::from ?

Reply:
We'll do this in another branch, and comment has been added, so done.